src/backend/distributed/commands/multi_copy.c

```diff
@@ -3554,30 +3567,53 @@ CopyGetPlacementConnection(HTAB *connectionStateHash, ShardPlacement *placement,
 		ConnectionStateListToNode(connectionStateHash, nodeName, nodePort);
 	if (HasReachedAdaptiveExecutorPoolSize(copyConnectionStateList))
 	{
-		connection =
-			GetLeastUtilisedCopyConnection(copyConnectionStateList, nodeName, nodePort);
-
 		/*
 		 * If we've already reached the executor pool size, there should be at
 		 * least one connection to any given node.
 		 */
-		Assert(connection != NULL);
+		connection = GetLeastUtilisedCopyConnection(copyConnectionStateList,
+													nodeName,
+													nodePort);
+
+		/*
+		 * Errors are supposed to cause immediate aborts (i.e. we don't
+		 * want to/can't invalidate placements), mark the connection as
+		 * critical so later errors cause failures.
+		 */
+		MarkRemoteTransactionCritical(connection);
 
 		return connection;
 	}
 
 	/*
-	 * For placements that haven't been assigned a connection by a previous command
-	 * in the current transaction, we use a separate connection per placement for
-	 * hash-distributed tables in order to get the maximum performance.
+	 * Enforce the requirements for adaptive connection connection
+	 * management (a.k.a., throttle connections if citus.max_shared_pool_size
+	 * reached)
 	 */
-	if (placement->partitionMethod == DISTRIBUTE_BY_HASH &&
-		MultiShardConnectionType != SEQUENTIAL_CONNECTION)
-	{
-		connectionFlags |= CONNECTION_PER_PLACEMENT;
-	}
+	int adaptiveConnectionManagementFlag =
+		AdaptiveConnectionManagementFlag(list_length(copyConnectionStateList));
+	connectionFlags |= adaptiveConnectionManagementFlag;
 
 	connection = GetPlacementConnection(connectionFlags, placement, nodeUser);
+	if (connection == NULL)
+	{
+		/*
+		 * The connection manager throttled any new connections, so pick an existing
+		 * connection with least utilization.
+		 */
+		connection = GetLeastUtilisedCopyConnection(copyConnectionStateList,
+													nodeName,
+													nodePort);
+
+		/*
+		 * Errors are supposed to cause immediate aborts (i.e. we don't
+		 * want to/can't invalidate placements), mark the connection as
+		 * critical so later errors cause failures.
+		 */
+		MarkRemoteTransactionCritical(connection);
```

Comment:
Given that it's in copyConnectionStateList, we should have already done this.

Reply:
I removed the line